### PR TITLE
AK: Add template specializations for `Optional<{,Fly}String>`

### PR DESCRIPTION
**Pull request by yyny**

### CODE REVIEW

**Comment by awesomekling** [comment]
Can we make the bit pattern for null be `0x0`?

Given this:
```cpp
class StringBase {
public:
    // Creates an empty (zero-length) String.
    constexpr StringBase()
        : StringBase(ShortString { .byte_count_and_short_string_flag = SHORT_STRING_FLAG })
    {
    }
```

The empty string is already `0x1`, so I don't think `0x0` would conflict with anything?

**Comment by yyny** [comment]
> Can we make the bit pattern for null be 0x0?

We _can_, but we would need an additional check in the destructor.

**Comment by awesomekling** [comment]
> > Can we make the bit pattern for null be 0x0?
> 
> We _can_, but we would need an additional check in the destructor.

If we're choosing between two evils, I would expect ref/unref on strings to be much more common then destroying strings. What do you think?

File: Userland/Libraries/LibTLS/Certificate.h [hunk 1]
```diff
@@ -218,7 +218,7 @@ class RelativeDistinguishedName {
     {
         auto entry = get(AttributeType::Cn);
         if (entry.has_value()) {
-            return entry.value();
```
Comment (awesomekling) [inline Userland/Libraries/LibTLS/Certificate.h:221]: I am confused by this change, “entry” is still live while we were invoking the copy constructor, no?

File: AK/Noncopyable.h [hunk 2]
```diff
@@ -25,3 +25,44 @@ public:                            \
...
+
+#define AK_MAKE_CONDITIONALLY_NONDESTRUCTIBLE(c, ...)   \
+    ~c()                                                \
+    requires(!(AK::Detail::IsDestructible __VA_ARGS__)) \
+    = delete
+
+#define AK_MAKE_CONDITIONALLY_DESTRUCTIBLE(c, ...)      \
+    ~c()                                                \
+    requires(!(AK::Detail::IsDestructible __VA_ARGS__)) \
+    = delete;                                           \
```
Comment (Hendiadyoin1) [inline AK/Noncopyable.h:67]: I do not like that these do not take care of `<>` and type packs on their own, but I can see that getting those to work might be a bit more involved

File: AK/Optional.h [hunk 3]
```diff
@@ -48,8 +49,135 @@ struct OptionalNone {
     explicit OptionalNone() = default;
 };
 
+template<typename T, typename Self>
```
Comment (Hendiadyoin1) [inline AK/Optional.h:52]:
```suggestion
template<typename T, typename Self = Optional<T>>
```
Should make the inheritance a bit nicer

**Review by Hendiadyoin1 — COMMENTED** [review]
Looks great. just a few nits on my side

File: Userland/Libraries/LibTLS/Certificate.h (hunk 1 above)
Comment (yyny) [reply Userland/Libraries/LibTLS/Certificate.h:221]: I was confused by this as well, and ChatGPT did _not_ help. But you are correct, this is not actually a bug in this code, but some kind of double free in the `Optional<String>`. And had I used strings with >7 characters I would have caught this in the tests... I will remove this change and fix the actual bug when I find the time.

File: AK/Optional.h (hunk 3 above)
Comment (Hendiadyoin1) [reply AK/Optional.h:52]: Also moving this commit up would be nice, as that removes a lot of "added then immediately removed code"

Comment (yyny) [reply AK/Optional.h:52]: I did it like this in case `OptionalBase` had to be reverted, but as I didn't see any objections to it, and since seems to work fine now, I can do this.

File: AK/Variant.h [hunk 4]
```diff
@@ -267,30 +267,9 @@ struct Variant
...
-    = delete;
-    Variant& operator=(Variant const&) = default;
-
-    Variant& operator=(Variant&&)
-    requires(!(IsMoveConstructible<Ts> && ...) || !(IsDestructible<Ts> && ...))
-    = delete;
-    Variant& operator=(Variant&&) = default;
+    AK_MAKE_CONDITIONALLY_COPYABLE(Variant, <Ts>&&...);
+    AK_MAKE_CONDITIONALLY_MOVABLE(Variant, <Ts>&&...);
+    AK_MAKE_CONDITIONALLY_DESTRUCTIBLE(Variant, <Ts>&&...);
```
Comment (Hendiadyoin1) [inline AK/Variant.h:272]: Technically an unrelated change

File: AK/Forward.h
Comment (Hendiadyoin1) [inline AK/Forward.h:1]: I dont see the changes to `NonnullRefPtr` mentioned in the commit message, or am I missing something

**Review by Hendiadyoin1 — COMMENTED** [review]
Two more nit's I found

**Comment by awesomekling** [comment]
Very awesome! Thank you for implementing this :)
